README.md
# AloFramework | Handlers #

A powerful error and exception handler of [AloFramework](https://github.com/aloframework/aloframework). It can automatically log errors and exceptions if an object implementing the [\Psr\Log\LoggerInterface](https://packagist.org/packages/psr/log) is passed on and will echo output based on your server's error reporting settings, altering its format depending on whether an error is raised during a HTTP or CLI call.

Latest release API documentation: [https://aloframework.github.io/handlers/](https://aloframework.github.io/handlers/)

[![License](https://poser.pugx.org/aloframework/handlers/license?format=plastic)](https://www.gnu.org/licenses/gpl-3.0.en.html)
[![Latest Stable Version](https://poser.pugx.org/aloframework/handlers/v/stable?format=plastic)](https://packagist.org/packages/aloframework/handlers)
[![Total Downloads](https://poser.pugx.org/aloframework/handlers/downloads?format=plastic)](https://packagist.org/packages/aloframework/handlers)

|                                                                                          dev-develop                                                                                          |                                                             Latest release                                                            |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|:-------------------------------------------------------------------------------------------------------------------------------------:|
| [![Dev Build Status](https://travis-ci.org/aloframework/handlers.svg?branch=develop)](https://travis-ci.org/aloframework/handlers)                                                            | [![Release Build Status](https://travis-ci.org/aloframework/handlers.svg?branch=1.4.1)](https://travis-ci.org/aloframework/handlers)  |
| [![SensioLabsInsight](https://insight.sensiolabs.com/projects/36b22482-e36a-44e3-a7de-ccf6e27999d1/mini.png)](https://insight.sensiolabs.com/projects/36b22482-e36a-44e3-a7de-ccf6e27999d1) | [![SensioLabsInsight](https://i.imgur.com/KygqLtf.png)](https://insight.sensiolabs.com/projects/36b22482-e36a-44e3-a7de-ccf6e27999d1) |

## Installation ##
Installation is available via Composer:

    composer require aloframework/handlers

## Usage ##

 - To enable the exception handler call `\AloFramework\Handlers\ExceptionHandler::register()`
 - To enable the error handler call `\AloFramework\Handlers\ErrorHandler::register()`
 - To enable the shutdown handler call `\AloFramework\Handlers\ShutdownHandler::register()`

### Logging ###
Every error and exception has to be logged in this package. You can supply your own logger to have more control; if you don't, [aloframework/log](https://github.com/aloframework/log) will be used with its default settings. 

## Configuration ##
Configuration is done via the classes in the `AloFramework\Handlers\Config` namespace. 

### Common ###

 - `CFG_CSS_PATH` - path to the CSS file which will style the HTML output. Defaults to **error.min.css** in the src directory.
 - `CFG_TRACE_MAX_DEPTH` - maximum number of debug backtrace items to display [**50**]
 - `CFG_BACKGROUND` - CLI output background colour [**default**]
 - `CFG_FOREGROUND_NOTICE` - CLI output notice level foreground colour [**cyan**]
 - `CFG_FOREGROUND_WARNING` - CLI output warning level foreground colour [**yellow**]
 - `CFG_FOREGROUND_ERROR` - CLI output error/exception level foreground colour [**red**]
 - `CFG_FORCE_HTML` - Whether to force HTML output even in CLI mode [**false**]
 - `CFG_REGISTER_SHUTDOWN_HANDLER` - **deprecated**, whether to register the shutdown handler when calling the deprecated `AbstractHandler::register()` method [**false**]

### Error Handlers' Config ###

 - `CFG_ERROR_LEVEL` - Which error levels to handle. Defaults to the value of `error_reporting()`.
 - `CFG_LOG_ERROR_LOCATION` - Whether to include the error location in the log [**true**]

### Exception Handlers' Config ###

 - `CFG_EXCEPTION_DEPTH` - Maximum number previous exceptions to output in the exception handler [**10**]
 - `CFG_LOG_EXCEPTION_LOCATION` Whether to include the exception location in the log [**true**]

## Sample Output ###

| **HTML**                                                                                              | **CLI**                                                                                              |
|-------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------|
| [Error](https://github.com/aloframework/handlers/blob/master/output-examples/error-html.png)         | [Error](https://github.com/aloframework/handlers/blob/master/output-examples/error-cli.png)         |
| [Exception](https://github.com/aloframework/handlers/blob/master/output-examples/exception-html.png) | [Exception](https://github.com/aloframework/handlers/blob/master/output-examples/exception-cli.png) |
| [Warning](https://github.com/aloframework/handlers/blob/master/output-examples/warning-html.png)     | [Warning](https://github.com/aloframework/handlers/blob/master/output-examples/warning-cli.png)     |
| [Notice](https://github.com/aloframework/handlers/blob/master/output-examples/notice-html.png)       | [Notice](https://github.com/aloframework/handlers/blob/master/output-examples/notice-cli.png)       |


CHANGELOG.md
# 1.4.1 #

@property PHPDocs added to the config classes, config keys' string identifiers relabeled

# 1.4 #

 - Configuration is now expected to be done via the [Configurable interface](https://github.com/aloframework/config),
  the old, constant-based configuration is now deprecated and will be removed in 2.0
 - Better tests' code coverage
 - Output examples added

# 1.3 #

Bugfixes:

 - ExceptionHandler now correctly assigns the `$lastRegisteredHandler` and `$lastReported` values

Error class:

 - Now has a static `$map` property, allowing easy text representation of error codes
 - Now has a shouldBeReported() static method
 - It's now possible to force HTML error output when in CLI mode by defining `define('ALO_HANDLERS_FORCE_HTML', true);`

ShutdownHandler:

 - Doesn't force the `E_CORE_ERROR` error code now
 - Has the `getLastRegisteredHandler` static method now

ErrorHandler:

 - Error labels are more detailed now (e.g. `CORE ERROR` instead of just `ERROR`)

Misc:

 - .gitattributes added
 - More extensive testing

# 1.2.1 #

Fixed the if statement in the shutdown handler, which should now allow it to handle fatal errors if no errors were reported beforehand.

# 1.2 #

 - The last reported error is tracked in ErrorHandler and can be retrived via the error handler's static 
 `getLastReportedError` method.
 - The last registered handler can now be retrieved via the static `getLastRegisteredHandler` method.
	 - Equivalent operations are available to the ExceptionHandler via `getLastReportedException` and `getLastRegisteredHandler`.
 - `E_CORE_ERROR` is now labelled as **FATAL ERROR**.
 - `E_CORE_ERROR` and `E_CORE_WARNING` are now tracked.
 - The shutdown handler is now available in the `ShutdownHandler` class and can be registered via `\AloFramework\Handlers\ShutdownHandler::register()` static method.
	 - By default it is disabled and won't be registered via AbstractHandler. You can enable this by defining `define('ALO_HANDLERS_REGISTER_SHUTDOWN', true)` before calling composer's autoloader.
 - `aloframework/log` is now a dependency - all errors should be tracked.
 

# 1.1 #

 - The `log` method in `ErrorHandler` now accepts two more optional arguments: `$file` & `$line`. If they are provided, the file and line where the error occurred will be appended to the log message.
 - The `log` method in `ExceptionHandler` now accepts an additional boolean parameter, `$includeLocation`, which defaults to `true`. Unless set to false, the exception file and line will be appended to the log message.
	 - The above two functionalities can be switched off by including the following two lines in your code before initialising composer: 
		 - `define('ALO_HANDLERS_LOG_ERROR_LOCATION', false);`
		 - `define('ALO_HANDLERS_LOG_EXCEPTION_LOCATION', false);`
	 - The above dual configuration is in place to allow easy conditional location logging, i.e. you can simply extend the handler class and overwrite the logging methods to only log under specific conditions instead of turning the functionality off entirely.
 - The default configuration file is only loaded when the class file gets included
 - General cleanup of unnecessary files
 - Code documentation is now a lot more detailed

# 1.0.5 #
Removed private props & methods from docs. Removed some unused class constants.

# 1.0.4 #
Updated how the handlers' `register()` methods work. They now determine the class via `get_called_class()`, allowing you to extend the handlers without having to overwrite the `register()` method.


src/config.default.php
<?php

    //    if (!defined('ALO_HANDLERS_CSS_PATH')) {
    //        /** Path to the handlers' CSS file for injection */
    //        define('ALO_HANDLERS_CSS_PATH', __DIR__ . DIRECTORY_SEPARATOR . 'error.min.css');
    //    }

    //    if (!defined('ALO_HANDLERS_ERROR_LEVEL')) {
    //        /** The error level to register the error handler for */
    //        define('ALO_HANDLERS_ERROR_LEVEL', error_reporting());
    //    }

    if (!defined('ALO_HANDLERS_EXCEPTION_DEPTH')) {
        /** Maximum number of previous exceptions to output */
        define('ALO_HANDLERS_EXCEPTION_DEPTH', 10);
    }

    //    if (!defined('ALO_HANDLERS_TRACE_MAX_DEPTH')) {
    //        /** Maximum number of items to appear in the debug backtrace */
    //        define('ALO_HANDLERS_TRACE_MAX_DEPTH', 50);
    //    }
    //
    //    if (!defined('ALO_HANDLERS_BACKGROUND')) {
    //        /** CLI output background colour */
    //        define('ALO_HANDLERS_BACKGROUND', 'default');
    //    }
    //
    //    if (!defined('ALO_HANDLERS_FOREGROUND_NOTICE')) {
    //        /** CLI output foreground colour for notices */
    //        define('ALO_HANDLERS_FOREGROUND_NOTICE', 'cyan');
    //    }
    //
    //    if (!defined('ALO_HANDLERS_FOREGROUND_WARNING')) {
    //        /** CLI output foreground colour for warnings */
    //        define('ALO_HANDLERS_FOREGROUND_WARNING', 'yellow');
    //    }
    //
    //    if (!defined('ALO_HANDLERS_FOREGROUND_ERROR')) {
    //        /** CLI output foreground colour for errors */
    //        define('ALO_HANDLERS_FOREGROUND_ERROR', 'red');
    //    }

    //    if (!defined('ALO_HANDLERS_LOG_ERROR_LOCATION')) {
    //        /**
    //         * Whether to include the error location in the log
    //         * @since 1.1
    //         */
    //        define('ALO_HANDLERS_LOG_ERROR_LOCATION', true);
    //    }

//    if (!defined('ALO_HANDLERS_LOG_EXCEPTION_LOCATION')) {
//        /**
//         * Whether to include the exception location in the log
//         * @since 1.1
//         */
//        define('ALO_HANDLERS_LOG_EXCEPTION_LOCATION', true);
//    }

    //    if (!defined('ALO_HANDLERS_REGISTER_SHUTDOWN')) {
    //        /** Whether to register the shutdown handler */
    //        define('ALO_HANDLERS_REGISTER_SHUTDOWN', false);
    //    }

    //    if (!defined('ALO_HANDLERS_FORCE_HTML')) {
    //        /** Whether to force HTML output */
    //        define('ALO_HANDLERS_FORCE_HTML', false);
    //    }
